processor/schemaprocessor/internal/migrate/attributes_test.go

```diff
@@ -0,0 +1,280 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migrate
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+type testTypeDefKey string
+type testTypeDefValue string
+
+func testHelperBuildMap(opts ...func(pcommon.Map)) pcommon.Map {
+	m := pcommon.NewMap()
+	for _, opt := range opts {
+		opt(m)
+	}
+	return m
+}
+
+func TestNewAttributeChangeSet(t *testing.T) {
+	t.Parallel()
+
+	t.Run("map of strings", func(t *testing.T) {
+		t.Parallel()
+
+		acs := NewAttributes(map[string]string{
+			"hello": "world",
+		})
+
+		expect := &AttributeChangeSet{
+			updates: map[string]string{
+				"hello": "world",
+			},
+			rollback: map[string]string{
+				"world": "hello",
+			},
+		}
+
+		assert.Equal(t, expect, acs, "Must match the expected value")
+	})
+
+	t.Run("typedef string values", func(t *testing.T) {
+		t.Parallel()
+
+		acs := NewAttributes(map[testTypeDefKey]testTypeDefValue{
+			"awesome": "awesome.service",
+		})
+
+		expect := &AttributeChangeSet{
+			updates: map[string]string{
+				"awesome": "awesome.service",
+			},
+			rollback: map[string]string{
+				"awesome.service": "awesome",
+			},
+		}
+
+		assert.Equal(t, expect, acs, "Must match the expected value")
+	})
+}
+
+func TestAttributeChangeSetApply(t *testing.T) {
+	t.Parallel()
+
+	for _, tc := range []struct {
+		name   string
+		acs    *AttributeChangeSet
+		attrs  pcommon.Map
+		expect pcommon.Map
+	}{
+		{
+			name: "no modifications",
+			acs:  NewAttributes(map[string]string{}),
+			attrs: testHelperBuildMap(func(m pcommon.Map) {
+				m.PutInt("test.cases", 1)
+			}),
+			expect: testHelperBuildMap(func(m pcommon.Map) {
+				m.PutInt("test.cases", 1)
+			}),
+		},
+		{
+			name: "Apply changes",
+			acs: NewAttributes(map[string]string{
+				"service_version": "service.version",
+			}),
+			attrs: testHelperBuildMap(func(m pcommon.Map) {
+				m.PutStr("service_version", "v0.0.1")
+			}),
+			expect: testHelperBuildMap(func(m pcommon.Map) {
+				m.PutStr("service.version", "v0.0.1")
+			}),
+		},
+		// Due to the way `pcommon.Map.RemoveIf` works,
+		// it forces both values to be removed from the the attributes.
```

Comment:
This appears to rely on undocumented behavior of RemoveIf. I think we either need to document it or change our implementation not to rely on undocumented behavior.

Reply:
Can we add a TODO to fix this?